### PR TITLE
Diff operation implementation

### LINKED ISSUE #138: diff command for Topology Templates
## Description
In this issue we describe an enhancement for opera to compare the existing Topology Template with a new one. Details are described in issue #130
## Current behaviour
Currently, opera does not provide a way to compare Topology Template. 
## Expected results
The user is able to run `diff` command providing a changed blueprint and inputs in a folder containing existing Service Template that was deployed before. The output is a human readable representation of templates differences, is formatted either as JSON or YAML and can be optionally saved in a file if `--output` option is provided.

### PR DESCRIPTION
**Pull request by alexmaslenn**
Closes #138 and discussed in #130

### CODE REVIEW

File: src/opera/commands/diff.py
Comment (anzoman) [inline src/opera/commands/diff.py:215]: Does this compare instances? Would then be better `topology_comparer.compare_instances`?

Also I tried to deploy `service.yaml` from `examples/hello/` and then I changed service template by adding one new input (that is not used anywhere):

```yaml
topology_template:
  inputs:
    marker:
      type: string
      default: default-marker
    input:
      type: string
      default: input
```

I also created `inputs.yaml`:

```yaml
marker: test
input: test-input
```

Then I executed `opera diff` and got the following result:

```console
(.venv) anzoman@ubuntu:~/Desktop/xopera-opera$ opera diff -i examples/hello/inputs.yaml examples/hello/service.yaml --format yaml
nodes:
  hello:
    interfaces:
      Standard:
        operations:
          create:
            inputs:
              marker:
              - default-marker
              - test
          delete:
            inputs:
              marker:
              - default-marker
              - test
```

So, I the difference says that I changed the value for `marker` input, but it doesn't say that I added a new input. Is it meant to be that only used values are seen on diff? Or should there be one additional yaml keyname? Like this:

```yaml
inputs:
  test:
  - test-input
```

File: src/opera/compare/comparisons.py
Comment (anzoman) [inline src/opera/compare/comparisons.py:51]: Why is this special handling needed? Can we update opera parser so that this extra handling will not be necessary?

File: src/opera/compare/template_comparer.py
Comment (anzoman) [inline src/opera/compare/template_comparer.py:14]: You use words `old` and `new` in python file with diff opera command. Would it be better to use them here as well instead of 1 and 2?

File: src/opera/compare/template_comparer.py
Comment (anzoman) [inline src/opera/compare/template_comparer.py:119]: Why are attributes skipped?

File: src/opera/compare/template_comparer.py
Comment (anzoman) [inline src/opera/compare/template_comparer.py:98]: Add `@staticmethod` to all the methods that can be static and start methods that are used only within classes with underscores (_). Although Python does not have strong distinctions between “private” and “public” variables, you can still use underscores to indicate that some functions/variables are not really meant to be a part of the public interface of some class.

**Review by anzoman — CHANGES_REQUESTED** [review]
Thanks @alexmaslenn, this looks promising! I requested a few changes. Also, great that you included unit tests. I would really like to see one integration test as well, where `opera diff` would be used in action. You can add this test (after you rebase this branch onto master) to `tests/integration/cli_commands`  where a testing script that tests all opera CLI commands resides, or you can create a fresh new integration test folder.

File: src/opera/compare/template_comparer.py
Comment (alexmaslenn) [reply src/opera/compare/template_comparer.py:98]: I'm not a great fan of static methods as dealing with concurrency becomes less intuitive. Anyway the case I would prefer  instance methods in this case that I would like to provide an ability to override some comparer methods later on if needed, the same way it can be done with Storage class. I do not know whether making some of methods static and some not makes any sense.

File: src/opera/compare/template_comparer.py
Comment (alexmaslenn) [reply src/opera/compare/template_comparer.py:119]: Attributes are usually assigned values during deployment (e.g. IP addresses). If we compare lets say identical templates attributes would certainly differ, while technically the deployments would be the identical.

File: src/opera/commands/diff.py
Comment (alexmaslenn) [reply src/opera/commands/diff.py:215]: The idea that we do not compare all inputs, but only inputs that have been reflected in some node properties. Basically if you just define a new input that does not affect any node there is no need for deployment update.

File: src/opera/compare/template_comparer.py
Comment (alexmaslenn) [reply src/opera/compare/template_comparer.py:14]: Well we use `new` and `old` in diff because it is clear which one is old (the deployed one) and which one is new, Here the comparison is more generic, we do not care actually about the creation time, just the differences.

File: src/opera/compare/comparisons.py
Comment (alexmaslenn) [reply src/opera/compare/comparisons.py:51]: While I agree that should be fixed at some point (if there is no reason for this discrepancy) I do not think it should be done as a part of this PR, but rather in a separate issue.

**Comment by alexmaslenn** [comment]
@anzoman thank you very much for your review, you made some valuable inputs. This particular issue was a hard one, I've missed some important stuff on the way. I've updated the code according to some of your comments and provided answers to those I'm open to discuss.

File: src/opera/compare/comparisons.py
Comment (anzoman) [reply src/opera/compare/comparisons.py:51]: Ok, then it'd be great that you open a new issue just to mention that this is something to be fixed.

File: src/opera/commands/diff.py
Comment (anzoman) [reply src/opera/commands/diff.py:199]: Ok, this sounds reasonable. We could then add comparing all inputs in the future if we will really need that functionallity.

File: src/opera/compare/template_comparer.py
Comment (anzoman) [reply src/opera/compare/template_comparer.py:14]: Aha, this is then ok.

File: src/opera/compare/comparisons.py
Comment (alexmaslenn) [reply src/opera/compare/comparisons.py:51]: Will do

File: src/opera/compare/template_comparer.py
Comment (anzoman) [reply src/opera/compare/template_comparer.py:119]: This makes sense then and can be kept as it is.

File: src/opera/compare/template_comparer.py
Comment (anzoman) [reply src/opera/compare/template_comparer.py:98]: Ok, I agree that we can skip those `@staticmethod` annotations here. However, do take a look if you have any methods that are meant to be used just within certain classes and prefix them with underscores so that users will know that these methods shouldn't be touched ([example](https://github.com/xlab-si/xopera-opera/blob/master/src/opera/parser/tosca/csar.py#L155)).

File: src/opera/commands/diff.py
Comment (anzoman) [inline src/opera/commands/diff.py:31]: This can be updated to something like:

```suggestion
        help="Optional: YAML or JSON file with inputs "
             "that will be used along with the comparison",
```

File: src/opera/commands/outputs.py
Comment (anzoman) [inline src/opera/commands/outputs.py:41]: Great! You can now also remove unused `import json` and `import yaml` imports from this file.

File: src/opera/compare/template_comparer.py
Comment (anzoman) [inline src/opera/compare/template_comparer.py:6]: This imports section can be changed to:

```suggestion
import filecmp

from os import path

from .diff import Diff
from .comparisons import Comparison, ListComparison, MapComparison
from opera.error import DataError
```

to keep consistency with opera commands and files that split up the imports into three sections.

File: src/opera/utils.py
Comment (anzoman) [inline src/opera/utils.py:75]: Add a missing newline

File: tests/unit/opera/compare/conftest.py
Comment (anzoman) [inline tests/unit/opera/compare/conftest.py:1]: Nice unit tests! Have you thought of adding an integration test as well? It'd be cool to have an integration test where opera diff would be used in action. You can add this test (after you rebase this branch onto master) to `tests/integration/cli_commands` where a testing script that tests all opera CLI commands resides, or you can create a fresh new integration test folder.

**Review by anzoman — COMMENTED** [review]
@alexmaslenn good job! I'm adding some more comments. @sstanovnik if you find some time, you can take a look too just to find anything that can be improved.

File: src/opera/compare/template_comparer.py
Comment (alexmaslenn) [reply src/opera/compare/template_comparer.py]: Done

File: tests/integration/compare/inputs1.yaml
Comment (anzoman) [inline tests/integration/compare/inputs1.yaml:1]: add a newline

File: tests/integration/compare/inputs2.yaml
Comment (anzoman) [inline tests/integration/compare/inputs2.yaml:1]: add a newline

File: tests/integration/compare/runme.sh
Comment (anzoman) [inline tests/integration/compare/runme.sh:18]: Remove one line at the end of this file. Currently there are two.

File: tests/integration/compare/types.yaml
Comment (anzoman) [inline tests/integration/compare/types.yaml:57]: You can use 2 spaces for YAML files like it's done in the rest of opera. Also check and refactor YAML file strings in unit tests.

```suggestion
tosca_definitions_version: tosca_simple_yaml_1_3

capability_types:
  hello_cap:
    derived_from: tosca.capabilities.Root
    properties:
      test1:
        default: "1"
        type: string
      test2:
        default: "1"
        type: string
      test3:
        default: "1"
        type: string

node_types:
  hello_type:
    derived_from: tosca.nodes.SoftwareComponent
    properties:
      time:
        default: "1"
        type: string
      test_map:
        default:
        type: map
    capabilities:
      test:
        type: hello_cap
    interfaces:
      Standard:
        inputs:
          marker:
            default: { get_input: marker }
            type: string
          time:
            default: { get_property: [SELF, time] }
            type: string
        operations:
          create: playbooks/create.yaml
          delete: playbooks/delete.yaml

  hello_type_old:
    derived_from: hello_type
    properties:
      day:
        default: "1"
        type: string

  hello_type_new:
    derived_from: hello_type
    properties:
      day:
        default: "2"
        type: string
```

File: tests/integration/compare/template2.yaml
Comment (anzoman) [inline tests/integration/compare/template2.yaml:72]: Use 2 spaces for YAML files

```suggestion
tosca_definitions_version: tosca_simple_yaml_1_3

imports:
  - types.yaml

topology_template:
  inputs:
    marker:
      type: string
      default: default-marker2

  node_templates:
    my-workstation:
      type: tosca.nodes.Compute
      attributes:
        private_address: localhost
        public_address: localhost

    hello-1:
      type: hello_type
      properties:
        time: "1"
      artifacts:
        def_file:
          type: tosca.artifacts.File
          file: files/file1_2.yaml
      requirements:
        - host: my-workstation

    hello-2:
      type: hello_type_new
      properties:
        time: "0"
      artifacts:
        def_file:
          type: tosca.artifacts.File
          file: files/file2.yaml
      capabilities:
        test:
          properties:
            test1: "3"
            test2: "3"
      requirements:
        - dependency: hello-1
        - host: my-workstation

    hello-3:
      type: hello_type
      properties:
        time: "0"
      requirements:
        - host: my-workstation

    hello-5:
      type: hello_type
      properties:
        time: "0"
      requirements:
        - host: my-workstation

    hello-6:
      type: hello_type
      properties:
        time: "0"
      artifacts:
        def_file:
          type: tosca.artifacts.File
          file: files/file2.yaml
      requirements:
        - dependency: hello-2
        - host: my-workstation
```

File: tests/integration/compare/template1.yaml
Comment (anzoman) [inline tests/integration/compare/template1.yaml:76]: Use 2 spaces for YAML files:

```suggestion
tosca_definitions_version: tosca_simple_yaml_1_3

imports:
  - types.yaml

topology_template:
  inputs:
    marker:
      type: string
      default: default-marker

  node_templates:
    my-workstation:
      type: tosca.nodes.Compute
      attributes:
        private_address: localhost
        public_address: localhost

    hello-1:
      type: hello_type
      properties:
        time: "0"
      artifacts:
        def_file:
          type: tosca.artifacts.File
          file: files/file1_1.yaml
      capabilities:
        test:
          properties:
            test1: "2"
            test2: "2"
      requirements:
        - host: my-workstation

    hello-2:
      type: hello_type_old
      properties:
        time: "0"
      artifacts:
        def_file:
          type: tosca.artifacts.File
          file: files/file1_1.yaml
      capabilities:
        test:
          properties:
            test1: "2"
            test2: "2"
      requirements:
        - host: my-workstation

    hello-3:
      type: hello_type
      properties:
        time: "0"
      requirements:
        - host: my-workstation

    hello-4:
      type: hello_type
      properties:
        time: "0"
      requirements:
        - host: my-workstation

    hello-6:
      type: hello_type
      properties:
        time: "0"
      artifacts:
        def_file:
          type: tosca.artifacts.File
          file: files/file2.yaml
      requirements:
        - dependency: hello-1
        - host: my-workstation
```

**Review by anzoman — CHANGES_REQUESTED** [review]
@alexmaslenn here is my new review. Integration tests work nicely. I see that you added a lot of new commits. Before final merging keep in mind to squash them so that you will end up with two or three commit at the end.

**Comment by alexmaslenn** [comment]
> @alexmaslenn Before final merging keep in mind to squash them so that you will end up with two or three commits at the end. 

As far as I understand GitHub allows to squash commits into one automatically before merging the PR, right?

File: src/opera/compare/instance_comparer.py
Comment (alexmaslenn) [reply src/opera/compare/instance_comparer.py:66]: The structure of the comparisons algorithm is that all methods should have common signature. It is more intuitive even if context is unused.

**Comment by anzoman** [comment]
> > @alexmaslenn Before final merging keep in mind to squash them so that you will end up with two or three commits at the end.
> 
> @anzoman As far as I understand GitHub allows to squash commits into one automatically before merging the PR, right?

Yes I believe that we have that enabled.

**Review by anzoman — DISMISSED** [review]
Thanks @alexmaslenn, nice work! I did a quick check again and I believe that code is now better. Integration and unit tests tests go through and the local installation of the package also works. So the last thing we need to do before we click on green merge button is to squash the commits and add a proper commit message and description.

**Comment by anzoman** [comment]
@alexmaslenn it seems that there are some conflicts left from your rebase on master. Please resolve them first and then squash the commits.

**Review by anzoman — APPROVED** [review]
@alexmaslenn I think that conflicts are now resolved and I hope that everything is now okay so that we can merge this one.

**Pull request merged**
